lib/cdo/i18n_backend.rb

```diff
@@ -164,12 +164,69 @@ def translate(locale, key, options = ::I18n::EMPTY_HASH)
       end
     end
 
-    class SimpleBackend < ::I18n::Backend::Simple
+    module Plugins
       include SmartTranslate
       include MarkdownTranslate
       include SafeInterpolation
       include I18nStringUrlTrackerPlugin
     end
+
+    class SimpleBackend < ::I18n::Backend::Simple
+      include Plugins
+    end
+
+    class LazyLoadableBackend < ::I18n::Backend::LazyLoadable
+      include Plugins
+
+      LOCALES_MAPPING = YAML.load_file(CDO.dir('dashboard/config/locales.yml')).each_with_object({}) do |(k, v), locales|
+        locales[k.to_sym] = v.to_sym if v.is_a?(String)
+      end.freeze
+
+      class ::I18n::Backend::LocaleExtractor
+        def self.locale_from_path(path)
+          # Extracts the locale name from the path, like "en", "en-US", "haw", "haw-HI", etc.
+          path[/\b([a-z]{2,3}(?:-[A-Z]{2})?)\b\.\w+$/, 1]&.to_sym
+        end
+      end
+
+      def load_translations(*files)
+        loaded_files.merge(files.flatten) if lazy_load?
+        super
+      end
+
+      def reload!
+        @loaded_files = nil if lazy_load?
+        super
+      end
+
+      def eager_load!
+        # Ignores eager loading instead of raising an error if lazy loading is enabled
+        super unless lazy_load?
+      end
+
+      protected
+
+      def loaded_files
+        @loaded_files ||= Set.new
+      end
+
+      def filenames_for_current_locale
+        valid_locales = Set.new(::I18n.fallbacks[::I18n.locale])
+        valid_locales << LOCALES_MAPPING[::I18n.locale] # en: :'en-US'
+        valid_locales.compact_blank
+
+        (::I18n.load_path.flatten - loaded_files.to_a).select do |path|
+          path_locale = ::I18n::Backend::LocaleExtractor.locale_from_path(path)
+          path_locale.nil? || valid_locales.include?(path_locale)
+        end
+      end
+
+      # Prevents from raising an error when the locale in the filename and
+      # the root locale key in the file content mismatch,
+      # like when `en.yml` has a root locale key of `en-US:` instead of `en:`
+      def assert_file_named_correctly!(...)
```

Comment:
This error highlights files that won't load due to mismatched locales in their names and content.
But since we changed the class to load such files, the error isn't relevant to us anymore.

Reply:
Just to make sure I understand correctly: is this one also a problem specific to our fallback functionality? As in, it's a problem because when the locale is `de-DE`, some of the files this method will be looking at will be named `en-US`? It makes sense to me that it would be too difficult to update the existing logic to take that into account; or at least, more difficult than it's really worth.

However, the comment (and my own initial reading when I wasn't thinking about the possibility of fallbacks) makes it seem like it's just a problem with us sometimes using `en` and sometimes using `en-US`, and that _does_ seem to me like something that would be easy enough to update the existing logic to accommodate. If that's not the case, could you update the comment to specify? Thanks!

Reply:
The issue with the COD i18n files is that many of them include a full locale like `de-DE` in the filename (e.g. [blocks.de-DE.yml](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/config/locales/blocks.de-DE.yml)) but the root locale key is just `de` and it triggers the error.

Reply:
The method ensures that for current locales like `pt-PT`, only `pt-PT` translations are loaded. However, the i18n backend has been changed to support fallback locales (`pt`, `pt-BR`, `en-US`, etc.). Also, many `pt-PT` files contain `pt` translations and modifications have been made to the code to allow these translations to be loaded regardless. 
Therefore, the original check for loading exclusively the current locale translations is no longer necessary.

Reply:
Hmm, I'm still not understanding why the check is no longer necessary. I get that it's now a lot more complicated to determine whether or not a given locale makes sense for a given file, but I don't see how our new loading scheme makes it _unnecessary_. I think we'd still like to know if a file with `pt-PT` in the filename is trying to load translations for the locale key `de`, for example.

Why not extract the `valid_locales` logic in `filenames_for_current_locale` into a helper method, then reimplement `assert_file_named_correctly` to check that the loaded locale is one of the multiple valid locales? Something along these lines (but probably optimized a bit better for performance):

```ruby
# Get the set of valid locales for a given base locale code; ie, for a base of `pt`, we expect valid locales `pt-PT`, `pt-BR`
def valid_locales_for(base_locale)
  valid_locales = Set.new(::I18n.fallbacks[base_locale])
  valid_locales << LOCALES_MAPPING[base_locale] # en: :'en-US'
  return valid_locales.compact_blank
end

def filenames_for_current_locale
  # Excludes already loaded i18n files to prevent them from being reloaded during locale switching.
  # For example, for the "de-DE" locale, the i18n fallbacks are "de", "en", and "en-US".
  # Since "en-US" (and "en") is the default locale, its i18n files are already loaded.
  (::I18n.load_path.flatten - loaded_files.to_a).select do |path|
    path_locale = ::I18n::Backend::LocaleExtractor.locale_from_path(path)
    path_locale.nil? || valid_locales_for(::I18n.locale).include?(path_locale)
  end
end

# Checks if a filename is named in correspondence to the translations it loaded.
# The locale extracted from the path must be either the single locale loaded in
# the translations or one of the expected variations of the locale, taking into
# account both long locale codes and fallbacks
def assert_file_named_correctly!(file, translations)
  loaded_locales = translations.keys.map(&:to_sym)
  expected_locale = LocaleExtractor.locale_from_path(file)
  unexpected_locales = loaded_locales.reject do |locale|
    locale == expected_locale || valid_locales_for(locale).include?(expected_locale)
  end

  raise FilenameIncorrect.new(file, expected_locale, unexpected_locales) unless unexpected_locales.empty?
end
```